common/types/block.go

```diff
@@ -39,47 +40,107 @@ func (id BlockID) AsHash32() Hash32 {
+func GetEffectiveGenesis() LayerID {
+	return NewLayerID(atomic.LoadUint32(&effectiveGenesis))
 }
 
-// LayerID is a uint64 representing a layer number. It is zero-based.
-type LayerID uint64
+// NewLayerID creates LayerID from uint32.
+func NewLayerID(value uint32) LayerID {
+	return LayerID{Value: value}
+}
+
+// LayerID is representing a layer number. Zero value is safe to use, and means 0.
+// Internally it is a simple wrapper over uint32 and should be considered immutable
+// the same way as any integer.
+type LayerID struct {
+	Value uint32
+}
 
 // GetEpoch returns the epoch number of this LayerID.
 func (l LayerID) GetEpoch() EpochID {
-	return EpochID(uint64(l) / uint64(getLayersPerEpoch()))
+	return EpochID(l.Value / getLayersPerEpoch())
 }
 
-// GetEffectiveGenesis returns when actual blocks would be created
-func GetEffectiveGenesis() LayerID {
-	return LayerID(atomic.LoadInt32(&EffectiveGenesis))
+// Add layers to the layer. Panics on wraparound.
+func (l LayerID) Add(layers uint32) LayerID {
+	nl := l.Value + layers
+	if nl < l.Value {
+		panic("layer_id wraparound")
+	}
+	l.Value = nl
+	return l
+}
+
+// Sub layers from the layer. Panics on wraparound.
+func (l LayerID) Sub(layers uint32) LayerID {
+	if layers > l.Value {
+		panic("layer_id wraparound")
+	}
+	l.Value -= layers
+	return l
 }
 
-// Add returns the LayerID that's layers (the param passed into this method) after l (this LayerID).
-func (l LayerID) Add(layers uint16) LayerID {
-	return LayerID(uint64(l) + uint64(layers))
+// Mul layer by the layers. Panics on wraparound.
+func (l LayerID) Mul(layers uint32) LayerID {
+	if l.Value == 0 {
+		return l
+	}
+	nl := l.Value * layers
+	if nl/l.Value != layers {
+		panic("layer_id wraparound")
+	}
+	l.Value = nl
+	return l
 }
 
-// Uint64 returns the LayerID as a uint64.
-func (l LayerID) Uint64() uint64 {
-	return uint64(l)
+// Uint32 returns the LayerID as a uint32.
+func (l LayerID) Uint32() uint32 {
+	return l.Value
+}
+
+// Before returns true if this layer is lower than the other.
+func (l LayerID) Before(other LayerID) bool {
+	return l.Value < other.Value
+}
+
+// After returns true if this layer is higher than the other.
+func (l LayerID) After(other LayerID) bool {
+	return l.Value > other.Value
+}
+
+// Duration returns different between current and other layer.
+func (l LayerID) Duration(other LayerID) uint32 {
+	if other.Value > l.Value {
+		panic(fmt.Sprintf("other (%d) must be before or equal to this layer (%d)", other.Value, l.Value))
+	}
+	return l.Value - other.Value
 }
 
 // Field returns a log field. Implements the LoggableField interface.
-func (l LayerID) Field() log.Field { return log.Uint64("layer_id", uint64(l)) }
+func (l LayerID) Field() log.Field { return log.Uint32("layer_id", l.Value) }
+
+// String returns string representation of the layer id numeric value.
+func (l LayerID) String() string {
+	return strconv.FormatUint(uint64(l.Value), 10)
```

Comment:
Why convert to uint64 here?

Reply:
strconv doesn't provide an alternative function that accepts uint32. I can format with fmt.Sprintf but it converts to uint64 internally too